main.go

```diff
@@ -80,94 +143,204 @@ func main() {
 		var (
-			body          []byte
-			err           error
-			verfier       *http.Cookie
-			verifierValue string
-			prettyJSON    bytes.Buffer
-
-			// The request will contain this code to exchange it for an access token.
-			code = request.URL.Query().Get("code")
+			res *http.Response
+			err error
 		)
-
-		if *pkceEnabled {
-			if verfier, err = request.Cookie("verifier"); err != nil {
-				log.Printf("%v\n", err)
-				return
-			}
-
-			if err = secureCookie.Decode("verifier", verfier.Value, &verifierValue); err != nil {
-				log.Printf("%v\n", err)
-				return
-			}
+		if token.AccessToken == "" {
+			templ.ExecuteTemplate(w, "error", token)
+			return
 		}
 
-		// Exchange code for an access token, include code verifier to validate it against challenge in ACP.
-		if body, err = client.Exchange(code, verifierValue); err != nil {
-			log.Printf("%v\n", err)
+		if res, err = fetchResource(client); err != nil {
+			log.Println(err)
 			return
 		}
 
-		if err = json.Indent(&prettyJSON, body, "", "\t"); err != nil {
-			log.Printf("error while decoding successful body response: %v\n", err)
-			return
+		templ.ExecuteTemplate(w, "balance", res)
+	}
+}
+
+func fetchResource(client acp.Client) (res *http.Response, err error) {
+	var (
+		httpClient *http.Client
+		req        *http.Request
+	)
+	if httpClient, err = newHTTPClient(client.Config); err != nil {
+		return nil, err
+	}
+
+	if req, err = newHTTPRequest(client.Config.ClientID); err != nil {
+		return nil, err
+	}
+
+	return httpClient.Do(req)
+}
+
+func newHTTPClient(c acp.Config) (*http.Client, error) {
+	var (
+		pool  *x509.CertPool
+		cert  tls.Certificate
+		certs = []tls.Certificate{}
+		data  []byte
+		err   error
+	)
+
+	if c.CertFile != "" && c.KeyFile != "" {
+		if cert, err = tls.LoadX509KeyPair(c.CertFile, c.KeyFile); err != nil {
+			return nil, fmt.Errorf("failed to read certificate and private key %v", err)
 		}
 
-		if _, err = fmt.Fprint(writer, prettyJSON.String()); err != nil {
-			log.Printf("error while writting successful body response: %v\n", err)
-			return
+		certs = append(certs, cert)
+	}
+
+	if pool, err = x509.SystemCertPool(); err != nil {
+		return nil, fmt.Errorf("failed to read system root CAs %v", err)
+	}
+
+	if c.RootCA != "" {
+		if data, err = os.ReadFile(c.RootCA); err != nil {
+			return nil, fmt.Errorf("failed to read http client root ca: %w", err)
 		}
+
+		pool.AppendCertsFromPEM(data)
+	}
+
+	return &http.Client{
+		Timeout: c.Timeout,
+		Transport: &http.Transport{
+			Proxy: http.ProxyFromEnvironment,
+			DialContext: (&net.Dialer{
+				Timeout:   30 * time.Second,
+				KeepAlive: 30 * time.Second,
+				DualStack: true,
+			}).DialContext,
+			MaxIdleConns:          100,
+			ResponseHeaderTimeout: c.Timeout,
+			IdleConnTimeout:       90 * time.Second,
+			TLSHandshakeTimeout:   10 * time.Second,
+			ExpectContinueTimeout: 1 * time.Second,
+			MaxIdleConnsPerHost:   runtime.GOMAXPROCS(0) + 1,
+			TLSClientConfig: &tls.Config{
```

Comment:
As this is just a sample app I'd set only Timeout and TLSClientConfig - we don't need a high-performance http client here.